### PR TITLE
trace: ftrace selftests

### PR DESCRIPTION
**Pull request by kalshett**
Added ftrace test cases from selftests

Signed-off-by: Kalpana Shetty <kalshett@in.ibm.com>

### CODE REVIEW

**Comment by kalshett** [comment]
@harish-24 @nasastry 
Incorporated suggested review comments. Can you review it once and take it for merging.

File: perf/trace.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            self.buldir = os.path.join(self.workdir, self.output)
+        os.chdir(os.path.join(self.buldir, self.testdir))
+
+    def test_ftrace(self):
+        '''
+        Execute ftrace tests
+        '''
+        self.args = self.params.get('args', default='')
+        self.log.info("ftrace tests...: %s" % self.args)
+        self.run_cmd_out(self, "./ftracetest -l /tmp ./test.d/%s" % self.args)
```
Comment (harish-24) [inline perf/trace.py:88]: `self` argument is not required to be passed here.

Comment (harish-24) [reply perf/trace.py:88]: instead of using `/tmp` as temp dir, please use `self.workdir`, or `self.teststmpdir` so that it gets cleaned up at the end of the test.

File: perf/trace.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            else:
+                git.get_repo(location, destination_dir=self.workdir)
+            self.buldir = os.path.join(self.workdir, self.output)
+        os.chdir(os.path.join(self.buldir, self.testdir))
+
+    def test_ftrace(self):
+        '''
+        Execute ftrace tests
+        '''
+        self.args = self.params.get('args', default='')
```
Comment (harish-24) [inline perf/trace.py:86]: Please add a default test here so that if the script is run without any yaml it still remains valid

File: perf/trace.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    Linux Trace tests
+    """
+
+    failed_tests = list()
+
+    @staticmethod
+    def run_cmd_out(self, cmd):
+        logfile = "/tmp/ftracetest.log"
+        process.system_output(cmd, ignore_status=True)
```
Comment (harish-24) [inline perf/trace.py:36]: Better to use `process.system` or `process.run` especially when the output is not required to process. Also better to check for its return code, now test will ERROR out when the rc is non-zero.

**Review by harish-24 — CHANGES_REQUESTED** [review]

File: perf/trace.py (hunk 1 above)
Comment (kalshett) [reply perf/trace.py]: Done.

File: perf/trace.py (hunk 2 above)
Comment (kalshett) [reply perf/trace.py]: Done.

File: perf/trace.py (hunk 3 above)
Comment (kalshett) [reply perf/trace.py]: Done.

File: perf/trace.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+class Linuxtrace(Test):
+
+    """
+    Linux Trace tests
+    """
+
+    failed_tests = list()
+
+    def run_cmd_out(self, cmd):
+        logfile = ("%s/ftracetest.log" % self.teststmpdir)
```
Comment (harish-24) [inline perf/trace.py:34]: braces are not necessary here

File: perf/trace.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        if process.system(cmd, shell=True):
```
Comment (harish-24) [inline perf/trace.py:35]: `ignore_status` is still required. otherwise the test still errors out.

**Review by harish-24 — CHANGES_REQUESTED** [review]

File: perf/trace.py (hunk 4 above)
Comment (kalshett) [reply perf/trace.py:34]: Done.

File: perf/trace.py (hunk 5 above)
Comment (kalshett) [reply perf/trace.py:35]: Take care of it.

**Comment by kalshett** [comment]
** SLES 15 SP3 runs status:
localhost:/home/kal/avocado-fvt-wrapper/tests/avocado-misc-tests/perf # cat x

avocado run trace.py -m /home/kal/avocado-fvt-wrapper/tests/avocado-misc-tests/perf/trace.py.data/trace.yaml

e.py.data/trace.yamle/kal/avocado-fvt-wrapper/tests/avocado-misc-tests/perf/trace
JOB ID     : 3b3fea7bc12173e4293fea08f4fbd396acdb85fd
JOB LOG    : /home/kal/avocado-fvt-wrapper/results/job-2021-03-25T03.43-3b3fea7/job.log
 (01/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-basic-version-distro-afa5:  PASS (21.31 s)
 (02/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-event-version-distro-ff65:  PASS (11.50 s)
 (03/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-ftrace-version-distro-f534:  FAIL: Failed Tests: ['[12] ftrace - Max stack tracer\t[FAIL]\n'] (50.43 s)
 (04/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-instances-version-distro-9ce5:  PASS (10.10 s)
 (05/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-trigger-version-distro-cc70:  PASS (75.30 s)
 (06/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-basic-version-upstream-1330:  PASS (65.56 s)
 (07/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-event-version-upstream-1d6d:  PASS (55.55 s)
 (08/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-ftrace-version-upstream-9541:  FAIL: Failed Tests: ['[13] ftrace - Max stack tracer\t[FAIL]\n'] (92.41 s)
 (09/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-instances-version-upstream-dd3d:  PASS (52.95 s)
 (10/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-trigger-version-upstream-c629:  FAIL: Failed Tests: ['[1] event trigger - test inter-event histogram trigger expected fail actions\t[XFAIL]\n', '[11] event trigger - test inter-event histogram trigger trace action with dynamic string param\t[FAIL]\n', '[13] event trigger - test synthetic_even... (153.99 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 3 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kal/avocado-fvt-wrapper/results/job-2021-03-25T03.43-3b3fea7/results.html
JOB TIME   : 590.66 s


** RHEL 8.4 runs status:
[root@ltcden9-lp4 perf]# cat x

avocado run trace.py -m /home/kal/avocado-fvt-wrapper/tests/avocado-misc-tests/perf/trace.py.data/trace.yaml
#avocado run hw_break.py -m /home/kal/avocado-fvt-wrapper/tests/avocado-misc-tests/perf/hw_break.py.data/hw_break.yaml

ests/avocado-misc-tests/perf/trace.py.data/trace.yamle/kal/avocado-fvt-wrapper/te
JOB ID     : d52760ded59347df49b33da47691142229bfa7ee
JOB LOG    : /home/kal/avocado-fvt-wrapper/results/job-2021-03-25T03.35-d52760d/job.log
 (01/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-basic-version-distro-afa5:  PASS (183.97 s)
 (02/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-event-version-distro-ff65:  PASS (128.86 s)
 (03/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-ftrace-version-distro-f534:  FAIL: Failed Tests: ['[5] ftrace - function pid filters\t[FAIL]\n'] (147.33 s)
 (04/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-instances-version-distro-9ce5:  PASS (89.38 s)
 (05/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-trigger-version-distro-cc70:  PASS (225.56 s)
 (06/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-basic-version-upstream-1330:  PASS (154.19 s)
 (07/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-event-version-upstream-1d6d:  PASS (52.85 s)
 (08/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-ftrace-version-upstream-9541:  FAIL: Failed Tests: ['[7] ftrace - function pid filters\t[FAIL]\n', '[8] ftrace - stacktrace filter command\t[FAIL]\n', '[13] ftrace - Max stack tracer\t[FAIL]\n', '[18] (instance)  ftrace - function pid filters\t[FAIL]\n', '[19] (instance)  ftrace - stacktrace... (91.55 s)
 (09/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-instances-version-upstream-dd3d:  PASS (40.70 s)
 (10/10) trace.py:Linuxtrace.test_ftrace;run-config-ftest_types-trigger-version-upstream-c629:  FAIL: Failed Tests: ['[11] event trigger - test inter-event histogram trigger trace action with dynamic string param\t[FAIL]\n', '[13] event trigger - test synthetic_events syntax parser\t[FAIL]\n'] (145.20 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 3 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kal/avocado-fvt-wrapper/results/job-2021-03-25T03.35-d52760d/results.html
JOB TIME   : 1260.76 s

**Comment by kalshett** [comment]
[job.rh.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/6203186/job.rh.log)
[job.sles.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/6203188/job.sles.log)

**Review by harish-24 — APPROVED** [review]
LGTM

**Comment by kalshett** [comment]
Thanks a lot @harish-24 for reviewing this test code. This helps.
